metrics/src/agent.rs

```diff
@@ -0,0 +1,266 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use crate::metrics::Metric;
+use crate::payload::construct_distribution_payload;
+
+use futures::lock::Mutex;
+use std::net::SocketAddr;
+use std::str;
+use std::sync::Arc;
+use std::time::UNIX_EPOCH;
+use std::time::{self, SystemTime};
+use tokio::net::UdpSocket;
+
+use hyper::header::CONTENT_TYPE;
+use hyper::http::HeaderValue;
+use hyper::{Body, Client, Request};
+use hyper_rustls::HttpsConnectorBuilder;
+
+const DOGSTATSD_PORT: u16 = 8125;
+const BUFFER_SIZE: usize = 8192;
+
+pub struct MetricsAgent {
+    config: Config,
+    buf: Arc<Mutex<Vec<Metric>>>,
+}
+
+impl MetricsAgent {
+    pub fn with_config(config: Config) -> Self {
+        Self {
+            config,
+            buf: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    pub async fn run(&self) {
+        // Create a UDP socket and bind to the dogstatsd port
+        let socket = UdpSocket::bind(
+            format!("0.0.0.0:{}", DOGSTATSD_PORT)
+                .parse::<SocketAddr>()
+                .unwrap(),
+        )
+        .await
+        .expect("Error binding to socket");
+
+        println!("Listening for dogstatsd packets on port {}", DOGSTATSD_PORT);
```

Comment:
we could use a logger for that? as this will be printed in customer stdout

Reply:
(I didn't see your comment before, feel free to ignore this)

Reply:
Ah yeah we should definitely introduce a logging library but I'd like to get _something_ in this week before I leave for vacation. I think it's ok for now to leave it as println, but if you feel strongly I can change it